src/ctia/lib/compojure/api/core.clj

```diff
@@ -48,7 +50,36 @@
   [middleware & body]
   `(core/middleware ~middleware ~@body))
 
-(defmacro context {:style/indent 2} [& args] `(core/context ~@args))
+(def ^:private allowed-context-options #{:tags :capabilities :description :return :summary})
```

Comment:
This is very difficult to really understand precisely what is going on. But I think the main problem comes from the auth middleware that set a `:lets`. By doing so, the `context` macro detects we have a `lets` and consider the `context` to be treated as `dynamic` (https://github.com/metosin/compojure-api/blob/master/src/compojure/api/meta.clj#L674) and this is AFAIU where the schema is potentially rebuilt.

Thus this put the `let` form just on top of the body and run for every request and not just at route init time.

So I am not sure this macro will be enough if you still allow the `:capabilities` param declaration at the context level.

Reply:
> But I think the main problem comes from the auth middleware that set a :lets.

Yes, this approach is incompatible with any middleware that sets `:lets`. It throws a compile-time error. That's why things like `:auth-identity` and `:identity-map` are pushed into the HTTP verbs.

> By doing so, the context macro detects we have a lets and consider the context to be treated as dynamic (https://github.com/metosin/compojure-api/blob/master/src/compojure/api/meta.clj#L674) and this is AFAIU where the schema is potentially rebuilt.

Ah I see. We use compojure-api 1.1.13 which doesn't have this optimization. The difference between my macro and compojure-api master is that mine throws a compile-time error on a dynamic route. 

> So I am not sure this macro will be enough if you still allow the :capabilities param declaration at the context level.

FWIW here's the expansion:

```clojure
(macroexpand
  `(context "" []
     :capabilities #{:foo :bar}
     (GET "" req {:status 200})))
;=>
(let* [routes__207013__auto__ (compojure.api.core/routes (ctia.lib.compojure.api.core/GET "" ctia.lib.compojure.api.core/req {:status 200}))
       capabilities207088 #{:bar :foo}]
  (compojure.api.core/context
    "" []
    :capabilities capabilities207088
    routes__207013__auto__))
```

Reply:
Perhaps even with the optimizations from compojure-api master, it would ok? This macro takes advantage of the fact that most arguments to `context` are expressions (as opposed to values like `:tags` and `:scopes`).